### ISSUE TITLE
getTiddlerRelinkReferences performance issue

### ISSUE THREAD

**Issue by linonetwo** [flibbles/tw5-relink#52]
On a large wiki (30k tiddlers) relink seems using too much time for index, block browser activity like typing.

![图片](https://github.com/flibbles/tw5-relink/assets/3746270/5d145aad-e160-4b8d-9d47-6ee0cec93290)

![图片](https://github.com/flibbles/tw5-relink/assets/3746270/826ed4aa-b38b-4946-a7ca-804fe226795e)

Here is the exported chrome devtool profile

[Trace-20240314T222352.json](https://github.com/flibbles/tw5-relink/files/14603618/Trace-20240314T222352.json)

And I'm using lazy-loading on huge wiki, so relink might not be useful on the browser side.

Can you make relink plugins only run on the nodejs side if it is on client-server setup? Like this example 

https://github.com/tiddly-gittly/tw-gamification/blob/e34cd9e7d604ba2c6b38b2ba39719c5f64073a2d/src/tw-gamification/reality-event-generator/change-filter/change-filter.ts#L17-L33

**Comment by flibbles** [comment]
Hey @linonetwo, good to see you again. [Remember this conversation?](https://github.com/Jermolene/TiddlyWiki5/issues/4767) Server-side relink won't ever become a thing because there's no interest in core in pursuing the support it'd need.

**Comment by flibbles** [comment]
But to answer your question about taking too much time, much of that is unavoidable. I've optimized where I can, but ultimately, Relink has to go through and analyze every single tiddler in order to have an idea of what references the current tiddler you're working with.

From your performance profiles, it at least looks like it's being good about only having to do that the first time. Once the Relink index is populated, it should be considerably faster.

**Comment by linonetwo** [comment]
The entrypoint of relink is https://github.com/flibbles/tw5-relink/blob/master/plugins/relink/js/bulkops.js

I'm considering adding a new param to startup, so on client-server mode, startup module with this param runs only on server-side, do you think that will solve the problem?

I think it will still be triggered by `tm-save-tiddler -> handleSaveTiddlerEvent -> this.wiki.relinkTiddler(draftOf,draftTitle)`, can you confirm is this the way your plugin is triggered? If so, I think run that startup module on server side will work.

**Comment by linonetwo** [comment]
Anyway, I'm consider use lazy-image for now, because lazy-all also break the backlink and backtransclude, need a one-directional way to sync indexer from server to client to fix this. (And only update indexer on the server, which can be done by let indexer-updater startup modules only run on server)

**Comment by flibbles** [comment]
Sorry for the late response. Busy week.

I mean, it's not impossible to make Relink do the work on the server. The path you describe is correct, I think. The `wiki.relinkTiddler` method is called, often by an action, but I think it actually comes from the `tm-relink-tiddler` event.

But I don't think that's relevant. Creating a lazy-compatible version of relink means making relink non-existent on the client, and changing the client to send different HTTP requests to the server. Right now, when a tiddler is renamed, it sends a PUT request to the server, and a DELETE request. It creates and removes. No rename. There'd have to be some kind of PUT or POST that initiates a relink on the server. This would mean a new `route` module.

Then, once the server has done that, it would send back a list of skinny tiddlers to the client to update its tiddler list.

This `route`, which would be something like `/recipes/default/tiddler/(tiddlerPath)`, and it might even be a PATCH request. This is something that would be more appropriate to exist in the core code, which is why I spoke to Jeremolene about it a while back, but he wasn't thrilled, even when I offered to do it.

But even then, there's also the relink Report mechanism, which is where you go to the TiddlerInfo tab to see links. This doesn't work unless the server does it as well, and that means a second new route. At least this one would be appropriate as a Relink feature only. But are you seeing why I decided not to pursue this? Especially after being told the lazy-tiddlywikis are an all-but-deprecated feature?
